pkg/metrics/prometheus.go

```diff
@@ -0,0 +1,152 @@
+package metrics
+
+import (
+	"github.com/grafana/grafana/pkg/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"net/http"
+	"strings"
+)
+
+const (
+	namespace = "grafana"
+	subsystem = "backend"
+)
+
+type PrometheusPublisher struct {
+	CounterCollectors map[string]*prometheus.CounterVec
+	GaugeCollectors   map[string]*prometheus.GaugeVec
+}
+
+func CreatePrometheusPublisher() (*PrometheusPublisher, error) {
+	go func() {
+		http.Handle("/metrics", prometheus.Handler())
+		err := http.ListenAndServe(":9200", nil)
```

Comment:
> So increase(logins.count[1d]) is the same as rate(logins.count[1d]) * 86400?

Yes, exactly the same.

> The question then is how is rate(logins.count[1d]) calculated? It would have to look at all samples for the whole day and derive an average rate for the whole day.

The short version is it subtracts the first value from the last, adjusts for counter resets and does some fudging to cover the gap between the first/last point and the time range.

> But given how step param works I do not think it does that.

The `query_range` API is semantically equivalent to multiple `query` API calls, each `step` seconds apart.

> Does Prometheus handle resets etc at ingest time?

All at query time.